python-server/pvCamera.py

```diff
@@ -0,0 +1,232 @@
+        return settingsList, imageArray_pv
+    except Exception as e:
+        await websocket.send_text(json.dumps({'error': str(e)}))
+        await websocket.close()
+        return None, None
+
+async def setup_signals(settingsList, websocket):
+    settingSignals = {}
+    try:
+        for item in settingsList:
+            signal = EpicsSignalRO(item['pv'], name=item['name'])
+            signal.get()
+            settingSignals[item['name']] = signal
+    except Exception as e:
+        await websocket.send_text(json.dumps({'error': str(e)}))
+        await websocket.close()
+        return False
+    return settingSignals
+
+async def check_connections(settingSignals, websocket):
+    for key, signal in settingSignals.items():
+        if not signal.connected:
+            print('setting PV ' + key + ' not connected')
+            await websocket.send_text(json.dumps({'error': f"{key} pv could not connect"}))
+            await websocket.close()
+            return False
+    return True
+
+async def check_signal_connection(signal, name, websocket):
+    if not signal.connected:
+        print('array data pv not connected, exiting')
+        await websocket.send_text(json.dumps({'error': f"The {name} pv could not connect"}))
+        await websocket.close()
+        return False
+    return True
+
+def update_enum_lists(settingSignals, colorModeEnumList, dataTypeEnumList):
+    #ColorMode and DataType should have an enum_strs attribute containing an array of
+    #strings whose index corresponds to the value held by the pv
+    #Overwrite the defaults with the actual values, if values don't exist than accept default
+    colorModeEnumList = getattr(settingSignals['colorMode'], 'enum_strs', colorModeEnumList)
+    dataTypeEnumList = getattr(settingSignals['dataType'], 'enum_strs', dataTypeEnumList)
+    return colorModeEnumList, dataTypeEnumList
+
+def update_dimensions(settingSignals, buffer):
+    colorModeValue = settingSignals['colorMode'].get()
+    dataTypeValue = settingSignals['dataType'].get()
+    tempDimensions = {
+        'x': settingSignals['sizeX'].get() - settingSignals['startX'].get(),
+        'y': settingSignals['sizeY'].get() - settingSignals['startY'].get(),
+        'colorMode': colorModeEnumList[colorModeValue],
+        'dataType': dataTypeEnumList[dataTypeValue]
+    }
+    buffer.put_nowait((None, time.time(), tempDimensions))
+
+# Main loop for streaming images
+
+async def handle_streaming(websocket, buffer):
+    try:
+        while True:
+            rawImageArray, timestamp, updatedSettings = await buffer.get()
+
+            if updatedSettings:
+                currentSettings = updatedSettings
+                await websocket.send_text(json.dumps(currentSettings))
+                continue
+
+            height, width, colorMode, dataType = currentSettings['y'], currentSettings['x'], currentSettings['colorMode'], currentSettings['dataType']
+
+            try:
```

Comment:
I would wrap array manipulation in its own non-async function then call it here with `asyncio.to_thread`. Numpy functions aren't async, but doing this allows multiple clients to run concurrently without being blocked.